Implement NodeIterator

Jinwoo-Song opened the pull request:
Implement NodeIterator's basic functionality. (Fixes #1235)  But the cases for node removals are not implemented yet. 

r? @jdm 
cc @yichoi

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/5981)

<!-- Reviewable:end -->

Jinwoo-Song committed:
Implement NodeIterator's attributes

 - root, whatToShow, and filter

Jinwoo-Song committed:
Implement more NodeIterator attribute and methods

Just implement the skelectons
 - referenceNode attribute
 - nextNode() and previousNode()

Jinwoo-Song committed:
Implement Document.createNodeIterator and rebase patch

Jinwoo-Song committed:
Implement nextNode(), previousNode() and detach() in NodeIterator

First patch to resolve #1235. We need more implementation for handling node removals,
and I'll implement in next patch.

highfive commented:
Thanks for the pull request, and welcome! The Servo team is excited to review your changes, and you should hear from @glennw (or someone else) soon.

nox commented on components/script/dom/nodeiterator.rs:
```diff
@@ -2,24 +2,308 @@
-    reflector_: Reflector
+    reflector_: Reflector,
+    root_node: JS<Node>,
+    reference_node: MutNullableHeap<JS<Node>>,
```
AFAICT you never set it to None, so MutHeap would be enough here.

larsbergstrom commented:
@jdm @Ms2ger Will one of the two of you be reviewing this?

nox commented:
<sup>**[components/script/dom/nodeiterator.rs, line 308 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHJ1H0Jquxj6D22in-)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L308)):</sup>
This can be confusing given we have dom::bindings::js::JS.

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 444 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L444)):</sup>
If #5902 gets merged before this PR, this test is going to pass.

---

<sup>**[tests/wpt/metadata/dom/traversal/NodeIterator.html.ini, line 5 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLHD3lxpkNYOuk3XM)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/traversal/NodeIterator.html.ini#L5)):</sup>
Why isn't that test passing?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
<sup>**[components/script/dom/nodeiterator.rs, line 97 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLersKhdTrSw8empk)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L97)):</sup>
No need for a helper if it is already a DOM method, remove prev_node().

---

<sup>**[components/script/dom/nodeiterator.rs, line 102 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLmf9aHeLOWqLDhGq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L102)):</sup>
Idem, remove next_node().

---

<sup>**[components/script/dom/nodeiterator.rs, line 118 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLYnDBMpuaMeFtquX)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L118)):</sup>
It is a bit confusing that the methods in the trait and the impl aren't in the same order.

---

<sup>**[components/script/dom/nodeiterator.rs, line 121 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHNMRDgC1vu-Qmlt61)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L121)):</sup>
To clean this helper and first_preceding_node_not_preceding_root, I'm pretty sure we can instead tweak TreeIterator in node.rs to take both a root and a current node.

---

<sup>**[components/script/dom/nodeiterator.rs, line 194 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHM6D7ODGQsuOXzRAw)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L194)):</sup>
Just write "Step 1.", "Step 2.", etc like in other files.

---

<sup>**[components/script/dom/nodeiterator.rs, line 234 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMmERCIjcrMNIAS1i)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L234)):</sup>
Link to spec.

---

<sup>**[components/script/dom/nodeiterator.rs, line 235 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMUfqpidLrphuia4y)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L235)):</sup>
Step 1.

---

<sup>**[components/script/dom/nodeiterator.rs, line 236 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMr2m2RAxWIrS9ruA)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L236)):</sup>
Root this right there.

---

<sup>**[components/script/dom/nodeiterator.rs, line 250 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMwtuN2eri30Hfrtt)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L250)):</sup>
Why do you need the ref forever?

---

<sup>**[components/script/dom/nodeiterator.rs, line 269 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHN1UbpiTxn7ffASdW)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L269)):</sup>
Why do you need the ref forever?

---

<sup>**[components/script/dom/nodeiterator.rs, line 282 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHN58JM5t4l93ISwrx)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L282)):</sup>
Why do you need the ref forever?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

jdm commented:
-S-awaiting-review +S-needs-code-changes 
@nox reviewed this with my blessing. Thanks for implementing this, @Jinwoo-Song!

---

Reviewed files:
- components/script/dom/document.rs @ r3
- components/script/dom/nodeiterator.rs @ r1, r2, r3, r4
- components/script/dom/webidls/Document.webidl @ r3
- components/script/dom/webidls/NodeIterator.webidl @ r1, r2, r4
- tests/wpt/metadata/dom/interfaces.html.ini @ r4
- tests/wpt/metadata/dom/traversal/NodeIterator.html.ini @ r4

---

<sup>**[components/script/dom/nodeiterator.rs, line 25 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp45dGleHKtpq1cpxny)** ([raw file](https://github.com/servo/servo/blob/410b038788951f1d3351c5862c7772e0c4145a42/components/script/dom/nodeiterator.rs#L25)):</sup>
nit: indent these to match `root_node`

---

<sup>**[components/script/dom/nodeiterator.rs, line 91 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp47HCRdQLX1f6bMQoR)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L91)):</sup>
Add a spec link.

---

<sup>**[components/script/dom/nodeiterator.rs, line 104 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp46NNyxCWjnJsLYRPV)** ([raw file](https://github.com/servo/servo/blob/732003ab5c90acecc872daa30f468587e5f04507/components/script/dom/nodeiterator.rs#L104)):</sup>
Let's just move the implementation of these into NextNode and PrevNode.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:umbrella: The latest upstream changes (presumably #6037) made this pull request unmergeable. Please resolve the merge conflicts.

Jinwoo-Song committed:
Update NodeIterator code to apply review comments.

Jinwoo-Song commented:
Review status: all files reviewed, 18 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/nodeiterator.rs, line 24 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JomCaYaxLICdfy3176n-r4-24)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L24)):</sup>
Done. Yes, latest spec says referenceNode cannot be NULL.

---

<sup>**[components/script/dom/nodeiterator.rs, line 33 \[r1\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp45dGleHKtpq1cpxny)** ([raw file](https://github.com/servo/servo/blob/410b038788951f1d3351c5862c7772e0c4145a42/components/script/dom/nodeiterator.rs#L33)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 91 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp47HCRdQLX1f6bMQoR)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L91)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 97 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLersKhdTrSw8empk)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L97)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 102 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLmf9aHeLOWqLDhGq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L102)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 118 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLYnDBMpuaMeFtquX)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L118)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 121 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHNMRDgC1vu-Qmlt61)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L121)):</sup>
Done. I added following_node() and preceding_nod() in node.rs and used them in nodeiterator.rs

---

<sup>**[components/script/dom/nodeiterator.rs, line 194 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHM6D7ODGQsuOXzRAw)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L194)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 219 \[r2\]](https://reviewable.io:443/reviews/servo/servo/5981#-Jp46NNyxCWjnJsLYRPV)** ([raw file](https://github.com/servo/servo/blob/732003ab5c90acecc872daa30f468587e5f04507/components/script/dom/nodeiterator.rs#L219)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 234 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMmERCIjcrMNIAS1i)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L234)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 235 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMUfqpidLrphuia4y)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L235)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 236 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMr2m2RAxWIrS9ruA)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L236)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 250 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHMwtuN2eri30Hfrtt)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L250)):</sup>
Done. Fixed the code not to use ref forever.

---

<sup>**[components/script/dom/nodeiterator.rs, line 269 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHN1UbpiTxn7ffASdW)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L269)):</sup>
Done. Fixed the code not to use ref forever.

---

<sup>**[components/script/dom/nodeiterator.rs, line 282 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHN58JM5t4l93ISwrx)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L282)):</sup>
Done. Fixed the code not to use ref forever.

---

<sup>**[components/script/dom/nodeiterator.rs, line 308 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHJ1H0Jquxj6D22in-)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L308)):</sup>
Would you suggest a candidate? :)

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 444 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L444)):</sup>
Great!

---

<sup>**[tests/wpt/metadata/dom/traversal/NodeIterator.html.ini, line 5 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLHD3lxpkNYOuk3XM)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/traversal/NodeIterator.html.ini#L5)):</sup>
Actually, I have no idea for this why this fails. Could you give me some advice?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 1 of 7 files reviewed, 10 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/document.rs, line 1318 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBpqj9p883M8UpxW-N)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/document.rs#L1318)):</sup>
https://dom.spec.whatwg.org/#dom-document-createnodeiteratorroot-whattoshow-filter

---

<sup>**[components/script/dom/node.rs, line 1306 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBsYyL29NtAOeyGH2k)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L1306)):</sup>
As a follow-up ticket to file, or if you want to do it right now, TreeIterator and FollowingRootIterator can be merged. Their only difference is that I stored the current depth instead of the root in the former.

Otherwise that's exactly what I had in mind.

---

<sup>**[components/script/dom/nodeiterator.rs, line 98 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLmf9aHeLOWqLDhGq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L98)):</sup>
Mmmh, that's a bit similar isn't? :) The point was to remove helper methods when they are used in a proper DOM method as-is. No need to have extra layers of indirections.

---

<sup>**[components/script/dom/nodeiterator.rs, line 103 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBqzusRLq3QA7CVHLl)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/nodeiterator.rs#L103)):</sup>
Ditto.

---

<sup>**[components/script/dom/nodeiterator.rs, line 157 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBtNJtY6KcqwIIHaBh)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/nodeiterator.rs#L157)):</sup>
if result == NodeFilterConstants::FILTER_ACCEPT { … }

---

<sup>**[components/script/dom/nodeiterator.rs, line 247 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHJ1H0Jquxj6D22in-)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L247)):</sup>
I don't have any good candidate. :s

Callback? Interpreted? NotNative? :P

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 357 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L357)):</sup>
In fact, it won't, I was wrong I fixed other lengths but not this one. Oh well. :)

---

<sup>**[tests/wpt/metadata/dom/traversal/NodeIterator.html.ini, line 5 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLHD3lxpkNYOuk3XM)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/traversal/NodeIterator.html.ini#L5)):</sup>
I'll look into it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 1 of 7 files reviewed, 11 unresolved discussions, all commit checks successful.

---

<sup>**[tests/wpt/metadata/dom/traversal/NodeIterator.html.ini, line 5 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLHD3lxpkNYOuk3XM)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/traversal/NodeIterator.html.ini#L5)):</sup>
So I've looked into it, and apparently it comes from callback::CallSetup::drop() not taking into account the CallSetup's exception handling mode. It shouldn't report the exception if the value is ExceptionHandling::Rethrow. You can either fix it or file a follow-up issue and I'll take care of it.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 1 of 7 files reviewed, 14 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/node.rs, line 790 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ0KOcwy37IcQLAPq)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L790)):</sup>
following_nodes

---

<sup>**[components/script/dom/node.rs, line 797 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ1lv9Ry4g-ljJVN5)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L797)):</sup>
preceding_nodes

---

<sup>**[components/script/dom/node.rs, line 804 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ3BJWYofgcu4zXvA)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L804)):</sup>
descending_last_children? I find that name confusing but I can't find a better one right now.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 2 of 7 files reviewed, 14 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/node.rs @ r5

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

Jinwoo-Song commented:
Review status: 2 of 7 files reviewed, 14 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/document.rs, line 1318 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBpqj9p883M8UpxW-N)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/document.rs#L1318)):</sup>
Done.

---

<sup>**[components/script/dom/node.rs, line 790 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ0KOcwy37IcQLAPq)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L790)):</sup>
Done.

---

<sup>**[components/script/dom/node.rs, line 797 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ1lv9Ry4g-ljJVN5)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L797)):</sup>
Done.

---

<sup>**[components/script/dom/node.rs, line 804 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqCQ3BJWYofgcu4zXvA)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L804)):</sup>
Done.

---

<sup>**[components/script/dom/node.rs, line 1306 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBsYyL29NtAOeyGH2k)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/node.rs#L1306)):</sup>
I'll follow up your advice in next patch. :)

---

<sup>**[components/script/dom/nodeiterator.rs, line 98 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLmf9aHeLOWqLDhGq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L98)):</sup>
Okay, I removed the helper function and moved the implementation in here.

---

<sup>**[components/script/dom/nodeiterator.rs, line 103 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBqzusRLq3QA7CVHLl)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/nodeiterator.rs#L103)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 157 \[r5\]](https://reviewable.io:443/reviews/servo/servo/5981#-JqBtNJtY6KcqwIIHaBh)** ([raw file](https://github.com/servo/servo/blob/e0726fe410090475626289b4e6a2993e20bb7a85/components/script/dom/nodeiterator.rs#L157)):</sup>
Done.

---

<sup>**[components/script/dom/nodeiterator.rs, line 247 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHJ1H0Jquxj6D22in-)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/components/script/dom/nodeiterator.rs#L247)):</sup>
I vote to 'Callback'. :)

---

<sup>**[tests/wpt/metadata/dom/traversal/NodeIterator.html.ini, line 5 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHLHD3lxpkNYOuk3XM)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/traversal/NodeIterator.html.ini#L5)):</sup>
I'll raise another issue for this, thanks!

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 2 of 7 files reviewed, 4 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/node.rs @ r6

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

Jinwoo-Song commented:
Review status: 2 of 7 files reviewed, 4 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/node.rs @ r6

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 444 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L444)):</sup>
It seems that all other tests for 'object length' fail. But I do not have idea for this right now.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

Manishearth commented:
Review status: 5 of 7 files reviewed, 4 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/webidls/Document.webidl @ r5
- components/script/dom/webidls/NodeIterator.webidl @ r5
- tests/wpt/metadata/dom/interfaces.html.ini @ r5

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 444 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L444)):</sup>
Try rebasing again and testing

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: all files reviewed, 4 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/document.rs @ r6
- components/script/dom/nodeiterator.rs @ r6
- components/script/dom/webidls/Document.webidl @ r5
- components/script/dom/webidls/NodeIterator.webidl @ r5
- tests/wpt/metadata/dom/interfaces.html.ini @ r5
- tests/wpt/metadata/dom/traversal/NodeIterator.html.ini @ r4

---

<sup>**[tests/wpt/metadata/dom/interfaces.html.ini, line 357 \[r4\]](https://reviewable.io:443/reviews/servo/servo/5981#-JpHL2OGsglT40fxp7jq)** ([raw file](https://github.com/servo/servo/blob/22983ffaa6419deb6c9cafdb2804bd49ed8db2b6/tests/wpt/metadata/dom/interfaces.html.ini#L357)):</sup>
@Manisheart Nah I was saying I thought I fixed them but I didn't, so even if he rebases nothing will change. I just made a remark for nothing.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

Manishearth commented:
@nox Is there anything left to resolve?

nox commented:
@Manishearth No, LGTM. :)

Manishearth commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit 6d570e9 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth 6d570e96de82f2da8519e30684e676482e135159 -->

bors-servo commented:
:hourglass: Testing commit 6d570e9 with merge 06e7b51...

bors-servo commented:
:broken_heart: Test failed - [mac1](http://build.servo.org/builders/mac1/builds/2066)

nox commented:
@Jinwoo-Song 

```
/html/dom/interfaces.html
-------------------------
PASS expected FAIL Document interface: document.implementation.createDocument(null, "", null) must inherit property "createNodeIterator" with the proper type (25)
PASS expected FAIL Document interface: calling createNodeIterator(Node,unsigned long,NodeFilter) on document.implementation.createDocument(null, "", null) with too few arguments must throw TypeError
```

Almost done! :)

Jinwoo-Song commented:
Updated test result!

nox commented:
Review status: all files reviewed, 4 unresolved discussions, all commit checks successful.
Reviewed files:
- components/script/dom/document.rs @ r7
- components/script/dom/node.rs @ r7
- tests/wpt/metadata/html/dom/interfaces.html.ini @ r7

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/5981)
<!-- Sent from Reviewable.io -->

Manishearth commented:
@bors-servo: r+

bors-servo commented:
:pushpin: Commit db25be7 has been approved by `Manishearth`

<!-- @bors-servo r=Manishearth db25be7bf61198999864559e302d96d860914abb -->

bors-servo commented:
:hourglass: Testing commit db25be7 with merge 2b52006...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/1868), [gonk](http://build.servo.org/builders/gonk/builds/1026), [linux1](http://build.servo.org/builders/linux1/builds/1859), [linux2](http://build.servo.org/builders/linux2/builds/1860), [mac1](http://build.servo.org/builders/mac1/builds/2067), [mac2](http://build.servo.org/builders/mac2/builds/2073)
